artemis-core-client/src/main/java/org/apache/activemq/artemis/utils/SoftValueLongObjectHashMap.java

```diff
@@ -21,26 +21,26 @@
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Comparator;
-import java.util.HashMap;
-import java.util.HashSet;
+import java.util.List;
 import java.util.Map;
 import java.util.Set;
 import java.util.TreeSet;
-import java.util.concurrent.atomic.AtomicLong;
 
+import io.netty.util.collection.LongObjectHashMap;
+import io.netty.util.collection.LongObjectMap;
 import org.jboss.logging.Logger;
 
-public class SoftValueHashMap<K, V extends SoftValueHashMap.ValueCache> implements Map<K, V> {
+public class SoftValueLongObjectHashMap<V extends SoftValueLongObjectHashMap.ValueCache> implements LongObjectMap<V> {
 
-   private static final Logger logger = Logger.getLogger(SoftValueHashMap.class);
+   private static final Logger logger = Logger.getLogger(SoftValueLongObjectHashMap.class);
 
    // The soft references that are already good.
    // too bad there's no way to override the queue method on ReferenceQueue, so I wouldn't need this
    private final ReferenceQueue<V> refQueue = new ReferenceQueue<>();
 
-   private final Map<K, AggregatedSoftReference> mapDelegate = new HashMap<>();
+   private final LongObjectMap<AggregatedSoftReference<V>> mapDelegate = new LongObjectHashMap<>();
```

Comment:
@michaelandrepearce Do you think it could be a problem to use it directly? I remember you've recently working with some primitive map of Netty...

Reply:
Why do you need to use it directly? They implement map interface. And the add primitive methods that will be invoked at runtime, you dont need to make the field the concrete impl. This is the trick all primitive impl use.

Reply:
But I wanted to avoid the boxing/unboxing cost: the page cache (that uses this soft value map) is sensible to GC cycles and the more is your allocation rate, the more likely you will have a minor GC, making the cache to clean some of its entries...
Avoiding boxing/unboxing is just a way to let the entries stays much longer and makes this map more effective.
TBH I don't know (the details I mean) why we are using a soft value map here: I would have used a LRU map instead with a fixed size of entries instead like https://github.com/real-logic/agrona/blob/master/agrona/src/main/java/org/agrona/collections/IntLruCache.java : need to ask @clebertsuconic about it...

Reply:
It dossnt box and unbox. Thats the trick they use. By implementing the same interface method but with primitive param at runtime the primitive version is invoked.

Reply:
Probably I've misunderstood your comment Michael, sorry.
I have understood that if you:
```
Map<Long, V> map;
...
V put(Long k, V value) {
    return map.put(k, value);
}

```
In this case a caller that will invoke put(...) is forced to provide a Long, instead of a primitive `long`...no? Or am I missing something?

Reply:
SoftValue is so it can gc. But if it doesnt has to can keep as much in memory as possible. Rather than a strict bound

Reply:
Youre missing something. You missing the wonder of the jvm at runtime. It like you have a method in interface that takes object. Your impl has to support that but it can implement the same name with a tighter typing and that will be invoked instead.

Reply:
That's clear to me, but let me explain better what I mean...
If we have:
```
private final LongObjectMap<AggregatedSoftReference<V>> mapDelegate
```
and externally we provide:
```
//(1)
V put(long k, V value) {
    return mapDelegate.put(k, value);
}

//(2)
V put(Long k, V value) {
    return mapDelegate.put(k, value);
}
```
If caller will call (1), the javac will hide a boxing (long -> Long) to allow calling `map.put(k, value)` given that 
mapDelegate has been declared as `Map<Long,...`.

Reply:
No it doesnt

Reply:
Try to debug it and you will tell me bud :)
Or look at the bytecode ;)

Reply:
So i think im missing something here, if i have two methods, the tighter typed one will get invoked.

Reply:
@michaelandrepearce 
Try to put a debug print on `Long::valueOf(long)` using:
```
   private static final Map<Long, Object> primitiveMap = new LongObjectHashMap<>();

   public static void main(String[] args) {
      final Object value = new Object();
      for (long l = 0 ; l<10000;l++) {
         final long x = l^l;
         primitiveMap.put(x, value);
      }
   }
```

Reply:
Ive realised what ive done, ive missread 

"LongObjectMap" as "LongObjectHashMap" ... you using LongObjectMap as field value not LongObjectHashMap.

Reply:
youre all good.